ambry-store/src/main/java/com/github/ambry/store/PersistentIndex.java

```diff
@@ -1532,6 +1544,50 @@ private long getAbsolutePositionInLogForOffset(Offset offset,
     return numPrecedingLogSegments * log.getSegmentCapacity() + offset.getOffset();
   }
 
+  /**
+   * @return the absolute read bytes represented by token.
+   * @throws StoreException
+   */
+  long getAbsoluteReadBytesFromIndexBasedToken(StoreFindToken indexBasedToken) throws StoreException {
+    return getAbsoluteReadBytesFromIndexBasedToken(indexBasedToken, validIndexSegments);
+  }
+
+  /**
+   * Get the absolute read bytes from index based token {@link StoreKey}.
+   * @param indexBasedToken the point until which the log has been read.
+   * @param indexSegments the list of index segments to use.
+   * @return the absolute read bytes represented by token.
+   * @throws StoreException
+   */
+  private long getAbsoluteReadBytesFromIndexBasedToken(StoreFindToken indexBasedToken, ConcurrentSkipListMap<Offset, IndexSegment> indexSegments)
+      throws StoreException {
+    Offset offset = indexBasedToken.getOffset();
+    LogSegment logSegment = log.getSegment(offset.getName());
+    if (logSegment == null || offset.getOffset() > logSegment.getEndOffset()) {
+      throw new IllegalArgumentException("Offset is invalid: " + offset + "; LogSegment: " + logSegment);
+    }
+    int numPrecedingLogSegments = getLogSegmentToIndexSegmentMapping(indexSegments).headMap(offset.getName()).size();
+    return numPrecedingLogSegments * log.getSegmentCapacity() + log.getSegmentCapacity() - getTotalOffsetAfterTokenPointedStoreKey(indexBasedToken);
```

Comment:
i don't think this is valid. we should probably do something like
```
numPrecedingLogSegments * log.getSegmentCapacity() + log.getEndOffset().getOffset() - getTotalOffsetAfterTokenPointedStoreKey(indexBasedToken);
```

Reply:
Thx for catching it! Fixed it.